docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,421 @@
+code omitted for brevity):
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       private DateTimeCodec $dateTimeCodec;
+
+       public function __construct()
+       {
+           $this->dateTimeCodec = new DateTimeCodec();
+       }
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->dateTimeCodec->decode($value->get('createdAt'));
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+               'createdAt' => $this->dateTimeCodec->encode($value->createdAt),
+           ]);
+       }
+   }
+
+Handling embedded documents
+---------------------------
+
+The previous example showed how to handle a single document. However, sometimes you want to handle fields that contain
+embedded documents. To show this, let's create an address document that we'll be embedding into our ``Person`` document.
+To ensure consistency, we're going to make this a readonly class:
+
+.. code-block:: php
+
+   <?php
+
+   final readonly class Address
+   {
+       public function __construct(
+           public string $street,
+           public string $postCode,
+           public string $city,
+           public string $country,
+       ) {
+   }
+
+We can now create a document codec for this class:
+
+.. code-block:: php
+
+   <?php
+
+   final class AddressCodec implements MongoDB\Codec\DocumentCodec
+   {
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           return new Address(
+               $value->get('street'),
+               $value->get('postCode'),
+               $value->get('city'),
+               $value->get('country'),
+           );
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               'street' => $value->street,
+               'postCode' => $value->postCode,
+               'city' => $value->city,
+               'country' => $value->country,
+           ]);
```

Comment:
Would performance be better without an intermediate array?
```suggestion
           $doc = new MongoDB\BSON\Document();
           $doc->set('street', $value->street);
           $doc->set('postCode, $value->postCode);
           $doc->set('city', $value->city);
           $doc->set('country', $value->country);
           
           return $doc;
```

Reply:
The BSON document class is currently read-only, so the only way to create one is through `fromPHP` or the other factory methods. The reason for this is that we only keep a `bson_t*` internally, which is essentially a byte stream of the BSON. In-place replacement only works for types with the same size, e.g. overwriting an int with another int. As soon as values grow or shrink in size, the BSON needs to be rewritten (i.e. create a new `bson_t*`, copy over everything except the property being changed, then write the changed property). Since the main motivation for adding the `Document` and `PackedArray` classes was to provide a way to access BSON data without decoding it to PHP objects, we refrained from adding a BSON writing mechanism. That said, there is PHPC-2172 which tracks creating an implementation of a BSON writer.